File: hs/src/Q400/Q448.hs
module Q400.Q448
(findDisappearedNumbersPlanA,
    findDisappearedNumbersPlanB
    ) where

import qualified Data.HashSet as HashSet
import qualified Data.List as List

findDisappearedNumbersPlanA :: [Integer] -> [Integer]
findDisappearedNumbersPlanA [] = []
findDisappearedNumbersPlanA n = filter (\x -> not (HashSet.member x fdn)) [1 .. len]
            where
                fdn = HashSet.fromList n
                len = toInteger . length $n


findDisappearedNumbersPlanB :: [Int] -> [Int]
findDisappearedNumbersPlanB [] = []
findDisappearedNumbersPlanB n = planb nl [1..(length n)]
            where
                nl = List.nub . List.sort $ n


planb :: [Int] -> [Int] -> [Int]
planb _ [] = []
planb [] y = y
planb (x@(xs:xss)) (y:ys)
             | xs == y = planb xss ys
             | xs > y  = y : planb x ys
             | otherwise = ys

-- [_] [_]
-- [_] (_:_:_)
-- (_:_:_) [_]
-- (_:_:_) (_:_:_)

-- n := len(nums)
-- for _, v := range nums {
--     v = (v - 1) % n
--     nums[v] += n
-- }
-- for i, v := range nums {
--     if v <= n {
--         ans = append(ans, i+1)
--     }
-- }
-- return
-- findDisappearedNumbersPlanC :: [Int] -> [Int]
-- findDisappearedNumbersPlanC [] = []
-- findDisappearedNumbersPlanC n = planc n (length n)
--
-- planc :: [Int] -> Int -> [Int]

File: golang/q400/q448.go
package q400

import (
	"sort"
)

func findDisappearedNumbersPlanB(nums []int) []int {
	if len(nums) == 0 {
		return nums
	}
	sort.Ints(nums)
	length := len(nums)
	length2 := length + 1
	i, j := 1, 0
	var result []int
	for j < length && i < length2 {
		if nums[j] == i {
			i++
			j++
		} else {
			for j < length && i < length2 && nums[j] < i {
				k := nums[j]
				j++
				for ; j < length && i < length2 && nums[j] == k; j++ {
					if k == i {
						i++
					}
				}
				for k = k + 1; k < i; k++ {
					result = append(result, k)
				}
			}
			for j < length && i < length2 && nums[j] > i {
				for ; nums[j] > i && i < length2; i++ {
					result = append(result, i)
				}
			}
		}
	}
	if i < length2 {
		k := nums[j-1] + 1
		for ; k < length2; k++ {
			result = append(result, k)
		}
	}
	return result
}

func findDisappearedNumbersPlanC(nums []int) []int {
	if len(nums) == 0 {
		return nums
	}
	var result []int
	for i := 0; i < len(nums); i++ {
		n := nums[i]
		for n > 0 {
			n, nums[n-1] = nums[n-1], -n
		}
		if nums[i] > 0 && nums[nums[i]-1] <= 0 {
			nums[i] = 0
		}
	}
	for i, v := range nums {
		if v >= 0 {
			result = append(result, i+1)
		}
	}
	return result
}
